xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xpn.xwiki.internal.doc;
+
+import java.util.ArrayList;
+
+import java.util.Collection;
+import java.util.ListIterator;
+import java.util.List;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.xpn.xwiki.doc.XWikiAttachment;
+
+/**
+ * AttachmentList that holds elements in order of filename.
+ * 
+ * @version $Id$
+ * @since 10.0RC1
+ */
+public abstract class XWikiAttachmentList extends ArrayList<XWikiAttachment>
```

Comment:
Shouldn't you extend AbstractNotifyOnUpdateList instead of reimplementing it ?

Reply:
Wouldn't it be more efficient to just create an AbstractNotifyOnUpdateList with a XWikiAttachmentList parameter like it was originally? I thought you were thinking of eliminating AbstractNotifyOnUpdateList - that's why I did it this way

Reply:
My point was to hide all this plumbing from XWikiDocument so that code in XWikiDocument is much simpler. It only deal with a XWikiAttachmentList which hide all the complexity.

Reply:
@tmortagne so are you saying I should extend it and then for method implementations call super and then modify the map as necessary?

Reply:
Now that I think about it, we can only extend one class and we are already extending ArrayList

Reply:
You don't really have to extend ArrayList (you just need it to implement List interface) and you don't need XWikiAttachmentList to take a parameter. Creating and passing an Arraylist to AbstractNotifyOnUpdateList can be a XWikiAttachmentList decision.

Reply:
I extended AbstractNotifyOnUpdateList on this PR https://github.com/xwiki/xwiki-platform/pull/695, that should be better since it implements list already